app/src/main/kotlin/compiler/intermediate_form/UniqueIdentifier.kt

```diff
@@ -0,0 +1,37 @@
+package compiler.intermediate_form
+
+class UnknownCharacter(message: String) : RuntimeException(message)
+
+class UniqueIdentifierFactory() {
+    private val knownConversionsToPlainASCII = mapOf(
+        'ą' to 'a',
+        'ć' to 'c',
+        'ę' to 'e',
+        'ł' to 'l',
+        'ó' to 'o',
+        'ś' to 's',
+        'ź' to 'z',
+        'ż' to 'z'
+    )
+
+    private val knownIdentifiers: MutableSet<String> = mutableSetOf()
+    fun build(prefix: String?, curr: String): UniqueIdentifier {
+        fun convertToPlainASCII(character: Char): Char {
+            if (character.code in 0..127) return character
+            if (character in knownConversionsToPlainASCII) return knownConversionsToPlainASCII[character]!!
+            throw UnknownCharacter("Unknown character: $character. Only polish signs are allowed outside of the plain ASCII characters.")
+        }
+        val identifierBuilder = StringBuilder()
+        prefix?.let {
+            identifierBuilder.append(it)
+            identifierBuilder.append('|')
+        }
+        curr.forEach { identifierBuilder.append(convertToPlainASCII(it)) }
+        var identifier = identifierBuilder.toString()
+        while (identifier in knownIdentifiers) identifier += "_"
```

Comment:
That method of resolving conflicts has one issue: after compilation (e.g. in debugger) one wouldn't be able to tell which function is which (technically, that is not true as it's all deterministic, but it would be non-intuitive and probably dependent on order of function declarations).
I would consider alternative scheme in which polish diacritic letter translates to multiple chars (something like 'zażółć' -> 'zaz#o#l#c#').